### PR TITLE
Support for browsers to play video streams

### PR DESCRIPTION
**Pull request by ehbello**
### Description
This pull request provides the following changes:

1) Install the missing video codecs for the Firefox browser.
2) Configure a dummy output for PulseAudio for all images to allow the consumption of sound data.
3) Add a new test to play video and validate the resulting images.

### Motivation and Context
These important changes allow browsers to reproduce and test sites with embedded video without memory leaks.

We, at [System73](https://system73.com), have been using Selenium to test our software for several years. We needed to implement these changes for Selenium nodes to work with video and now we are sharing them with the community.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
- [x] I have read the [contributing](https://selenium.dev/documentation/en/contributing/) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: Base/Dockerfile [hunk 1]
```diff
@@ -46,7 +46,7 @@ RUN echo "${TZ}" > /etc/timezone \
 RUN useradd seluser \
          --shell /bin/bash  \
          --create-home \
-  && usermod -a -G sudo seluser \
+  && usermod -a -G sudo,audio,video seluser \
```
Comment (diemol) [inline Base/Dockerfile]: Just out of curiosity, why are these permissions needed?

File: NodeBase/Dockerfile [hunk 2]
```diff
@@ -13,6 +13,7 @@ USER root
 RUN apt-get update -qqy \
   && apt-get -qqy install \
     xvfb \
+    pulseaudio \
```
Comment (diemol) [inline NodeBase/Dockerfile]: I can see that the video does not load in Firefox, but it does load in Chrome. Having said that, why is `pulseaudio` installed for both? Shouldn't it be only for Firefox?

File: NodeBase/start-selenium-node.sh [hunk 3]
```diff
@@ -7,6 +7,16 @@ if [ ! -e /opt/selenium/config.json ]; then
   exit 1
 fi
 
+# Start the pulseaudio server
+pulseaudio -D --exit-idle-time=-1
```
Comment (diemol) [inline NodeBase/start-selenium-node.sh]: Shouldn't this be done only for the Firefox node?

**Comment by maximehyh** [comment]
Hi there, 
Thanks for those updates, I tested the FF node and video streams are playing as expected. 
Any idea when/if this PR will be merged?

**Comment by diemol** [comment]
@maximehyh I gave some feedback to the PR creator, I hope they can address the comments, because the structure for Selenium 4 will be different, so if we do not merge this now, it will need to be done from scratch.

File: Base/Dockerfile (hunk 1 above)
Comment (ehbello) [reply Base/Dockerfile]: I assumed that **pulseaudio** needs audio permissions, but as I read in the documentation, it seems that it is not necessary.

The `video` group was added while trying to support hardware acceleration for browsers and is currently probably useless, so I forgot to remove it.

I am going to check if they are really necessary and I will come back again.

File: NodeBase/Dockerfile (hunk 2 above)
Comment (ehbello) [reply NodeBase/Dockerfile]: The video play is fixed by installing the `libavcodec-extra` package here, only for Firefox. https://github.com/SeleniumHQ/docker-selenium/pull/994/files#diff-a7327fbb68be844da4c26c10bc15c5c2R16

The installation of `pulseaudio` package for all browsers is because Chrome is also affected by the memory leak caused if there is no sound consumption.

You can check it. Simply load an HLS video site in either browser and wait. You could see a continuous increase in memory consumption until the OOM kill of the container.

File: NodeBase/start-selenium-node.sh (hunk 3 above)
Comment (ehbello) [reply NodeBase/start-selenium-node.sh]: According to my answer above https://github.com/SeleniumHQ/docker-selenium/pull/994#discussion_r380614139, `pulseaudio` daemon is required here because all browsers are affected by the memory leak caused if there is no sound consumption.

**Comment by ehbello** [comment]
@maximehyh, @diemol, thanks for your feedback. I am going to rebase the branch to update this PR and include the latest changes in `master`.

**Comment by ehbello** [comment]
@diemol, answering to your question about the docker image sizes, here is a summary:

```
REPOSITORY                          TAG                           IMAGE ID            CREATED             SIZE
selenium/standalone-opera-debug     3.141.59-zirconium-system73   3396aecb4a17        2 minutes ago       998MB
selenium/standalone-firefox-debug   3.141.59-zirconium-system73   c401efb0a638        2 minutes ago       864MB
selenium/standalone-chrome-debug    3.141.59-zirconium-system73   6d01805245ca        2 minutes ago       926MB
selenium/standalone-opera           3.141.59-zirconium-system73   c7dc2c99e169        2 minutes ago       967MB
selenium/standalone-chrome          3.141.59-zirconium-system73   2be9cf1dbf12        2 minutes ago       895MB
selenium/standalone-firefox         3.141.59-zirconium-system73   b46cae28ebc6        2 minutes ago       832MB
selenium/node-opera-debug           3.141.59-zirconium-system73   c9a3b6ffb7b7        2 minutes ago       998MB
selenium/node-firefox-debug         3.141.59-zirconium-system73   b354156c2bab        2 minutes ago       864MB
selenium/node-chrome-debug          3.141.59-zirconium-system73   c9dde7fa71f0        3 minutes ago       926MB
selenium/node-opera                 3.141.59-zirconium-system73   787f44ad700d        4 minutes ago       967MB
selenium/node-firefox               3.141.59-zirconium-system73   11549049e910        5 minutes ago       832MB
selenium/node-chrome                3.141.59-zirconium-system73   642aae28505d        6 minutes ago       895MB
selenium/node-base                  3.141.59-zirconium-system73   9f099b929fc4        7 minutes ago       542MB
selenium/hub                        3.141.59-zirconium-system73   a9bf1726b87a        9 minutes ago       263MB
selenium/base                       3.141.59-zirconium-system73   b672db5d829c        9 minutes ago       263MB
```

**Comment by ehbello** [comment]
@diemol, answering to your question about the docker image sizes, here is a summary:

```
REPOSITORY                          TAG                           IMAGE ID            CREATED             SIZE
selenium/standalone-opera-debug     3.141.59-zirconium-system73   3396aecb4a17        2 minutes ago       998MB
selenium/standalone-firefox-debug   3.141.59-zirconium-system73   c401efb0a638        2 minutes ago       864MB
selenium/standalone-chrome-debug    3.141.59-zirconium-system73   6d01805245ca        2 minutes ago       926MB
selenium/standalone-opera           3.141.59-zirconium-system73   c7dc2c99e169        2 minutes ago       967MB
selenium/standalone-chrome          3.141.59-zirconium-system73   2be9cf1dbf12        2 minutes ago       895MB
selenium/standalone-firefox         3.141.59-zirconium-system73   b46cae28ebc6        2 minutes ago       832MB
selenium/node-opera-debug           3.141.59-zirconium-system73   c9a3b6ffb7b7        2 minutes ago       998MB
selenium/node-firefox-debug         3.141.59-zirconium-system73   b354156c2bab        2 minutes ago       864MB
selenium/node-chrome-debug          3.141.59-zirconium-system73   c9dde7fa71f0        3 minutes ago       926MB
selenium/node-opera                 3.141.59-zirconium-system73   787f44ad700d        4 minutes ago       967MB
selenium/node-firefox               3.141.59-zirconium-system73   11549049e910        5 minutes ago       832MB
selenium/node-chrome                3.141.59-zirconium-system73   642aae28505d        6 minutes ago       895MB
selenium/node-base                  3.141.59-zirconium-system73   9f099b929fc4        7 minutes ago       542MB
selenium/hub                        3.141.59-zirconium-system73   a9bf1726b87a        9 minutes ago       263MB
selenium/base                       3.141.59-zirconium-system73   b672db5d829c        9 minutes ago       263MB
```

File: NodeBase/Dockerfile (hunk 2 above)
Comment (ddavison) [reply NodeBase/Dockerfile]: Let's revert this change also, since we don't modify these Dockerfile's directly :+1:

File: NodeFirefox/Dockerfile [hunk 4]
```diff
@@ -13,7 +13,7 @@ USER root
 ARG FIREFOX_VERSION=latest
 RUN FIREFOX_DOWNLOAD_URL=$(if [ $FIREFOX_VERSION = "latest" ] || [ $FIREFOX_VERSION = "nightly-latest" ] || [ $FIREFOX_VERSION = "devedition-latest" ]; then echo "https://download.mozilla.org/?product=firefox-$FIREFOX_VERSION-ssl&os=linux64&lang=en-US"; else echo "https://download-installer.cdn.mozilla.net/pub/firefox/releases/$FIREFOX_VERSION/linux-x86_64/en-US/firefox-$FIREFOX_VERSION.tar.bz2"; fi) \
   && apt-get update -qqy \
-  && apt-get -qqy --no-install-recommends install firefox \
+  && apt-get -qqy --no-install-recommends install firefox libavcodec-extra \
```
Comment (ddavison) [inline NodeFirefox/Dockerfile]: Should we revert this change since we don't modify the Dockerfile's directly?

File: tests/SeleniumTests/__init__.py [hunk 5]
```diff
@@ -39,6 +42,20 @@ def test_visit_basic_auth_secured_page(self):
         page_message = driver.find_element_by_css_selector('.example p').text
         self.assertTrue(page_message == 'Congratulations! You must have the proper credentials.')
 
+    def test_play_video(self):
+        driver = self.driver
+        driver.get('https://hls-js.netlify.com/demo/')
```
Comment (ddavison) [inline tests/SeleniumTests/__init__.py]: Should we propose an PR to [the-internet](https://github.com/tourdedave/the-internet) since Netlify isn't under the umbrella of the Selenium community?

Perhaps we could add a Video section [here](http://the-internet.herokuapp.com/).\

Thoughts?

File: NodeFirefox/Dockerfile (hunk 4 above)
Comment (ehbello) [reply NodeFirefox/Dockerfile]: Done.

File: NodeBase/Dockerfile (hunk 2 above)
Comment (ehbello) [reply NodeBase/Dockerfile]: @ddavison Done.

File: NodeBase/start-selenium-node.sh (hunk 3 above)
Comment (diemol) [reply NodeBase/start-selenium-node.sh]: Thanks for clarifying!

File: Base/Dockerfile (hunk 1 above)
Comment (diemol) [reply Base/Dockerfile]: I see you removed it, I assume it was not needed, thanks!

File: tests/SeleniumTests/__init__.py (hunk 5 above)
Comment (diemol) [reply tests/SeleniumTests/__init__.py]: I think we can tackle this in a different PR, and not as a prerequisite for this.
In general, we should try to come up with our own site that can be used for our tests and also for our docs.
I'll resolve this conversation, I hope that's ok.
